bpf_sock: Implements a new way to detect host netns on Kind.

chkhalt opened the pull request:
This commit follows the suggestion made by @borkmann, @brb and @aditighag with little modifications.

I really appreciate any suggestions.

Fixes: #14956

anfernee commented on pkg/datapath/loader/base.go:
```diff
@@ -255,6 +259,52 @@ func (l *Loader) Reinitialize(ctx context.Context, o datapath.BaseProgramOwner,
 		ret  error
 	)
 
```
The next 2 functions can be deduped. Looks like the only different thing is the first argument of `unix.Socket()`.

anfernee reviewed (commented):
Gave a first scan. It looks pretty good.

aditighag commented on bpf/bpf_sock.c:
```diff
@@ -446,6 +511,11 @@ __sock4_health_fwd(struct bpf_sock_addr *ctx __maybe_unused)
 int sock4_connect(struct bpf_sock_addr *ctx)
 {
+	if (sock_has_magic_cookie(ctx)) {
+		per_agent_hostns_set_cookie(ctx);
```
This will be executed on every `connect` call that `bpf_sock` intercepts for workloads, etc. AFAIU, I don't think that's the intention of @borkmann 's proposal. Netns cookies [remain constant](https://man7.org/linux/man-pages/man7/BPF-HELPERS.7.html) for their lifetime so this should be a one-time operation?

@borkmann I wonder if this workflow could work -  Agent creates a dummy socket, and passes down its socket cookie to `bpf_sock_dummy` program. It then makes a socket call that'll be intercepted and matched by the dummy program. The program then gets and stores (either writes to a map, or passes it down to the agent via ring buffer) the netns cookie (aka the "host netns" cookie) of this socket. These operations need to happen *before* `bpf_sock` is attached. The agent will then pass it to the `bpf_sock` program while loading. 
I think this is quite similar to your proposal, but I didn't quite understand couple of details around the ordering and such in your proposal.

chkhalt replied:
Good point! I will change the code by implementing a one-time operation. Thanks ;)

Commit:
bpf_sock: Implement a new way to detect host netns on Kind.

Sometimes cilium needs to know if a socket belongs to a host netns.
This is currently done by comparing the netns cookie (to which the socket belongs)
against the host netns.

But when cilium is running on kind this approach doesn't work, because
cilium-agent is running in a non-host netns.

This commit implements the suggestion made by Daniel Borkmann (with few
modifications) and defines a new per cilium-agent "host" netns, used when kind
is detected.

Fixes: cilium#14956

Signed-off-by: Joao Victorino <joao@accuknox.com>

borkmann commented on pkg/datapath/loader/base.go:
```diff
@@ -424,6 +543,9 @@ func (l *Loader) Reinitialize(ctx context.Context, o datapath.BaseProgramOwner,
 	}
 
+	// Required when cilium is running on Kind
+	args[initArgNetNsCookie] = getNetNsCookie(ctx)
```
Definitely a tricky issue. I discussed the overall approach with @brb a bit, and the problem I currently see in here is that it's racy. Meaning, whenever we Reinitialize() (one example would be agent up/down-grade), we shortly swap the bind() progs from the existing socket LB with the cookie probe ones. While they do SYS_REJECT, my concern is that it will lead to hard to debug application issues in production (why does my app occasionally return -EPERM on bind?). Our thinking is we expose the netns cookie via getsockopt similar to the socket cookie and route it as a stable fix. Once that is complete, we can just allocate a socket from the agent in getNetNsCookie(), do the getsockopt() call, and are done with it.

chkhalt replied:
> While they do SYS_REJECT, my concern is that it will lead to hard to debug application issues in production (why does my app occasionally return -EPERM on bind?).

That's an amazing catch! I agree ;)

> Our thinking is we expose the netns cookie via getsockopt similar to the socket cookie and route it as a stable fix. Once that is complete, we can just allocate a socket from the agent in getNetNsCookie(), do the getsockopt() call, and are done with it.

So, the plan could be to add a `BPF_PROG_TYPE_CGROUP_SOCKOPT` (on bpf_sock.c, for exemple) which ensures an unused argument (assuming `SO_NETNS_COOKIE=-1`, for exemple) and (maybe) `EINVAL` as syscall ret. And that, under these circumstances, set `optval=netns`?

Or would it be better wait for linux kernel to support netns cookie?
Thanks for the discussion!
